### DISCUSSION TITLE
Path Flag for .ini

### DISCUSSION THREAD

**Discussion by mgpat-gm** [diasurgical/DevilutionX#1336] [Q&A] [answered]
I mainly play versin 1.0.3. I would like to play on occasion whatever the recent release is with other players. I would like to run two different versions, using the same save dir. That way I do not have to move char files around.

Since the .ini is in the save dir, I would be bound to single .ini for both versions, if I direct each version to startup with the same save dir.

Are the .ini files compatible from old versions to new? I'm not concerned at this point with using a 1.1.0 .ini for 1.0.3. It would be great to have a launch flag for --ini-path since the game automatically generates the .ini.

If it automatically generated the .ini in the executable directory instead, you would have to run each executable in a different directory if you had different settings.

Ideally I'd have devilutionX-1.0.3 and devilutionX-x.x.x executables in the same folder. I would create a launch script or shortcut for whichever .ini/save folder/data dir I want to use. I have a different save folder and launch script for mules.

Examples:
/somepath/devilutionx-1.0.3 -n  --save-dir /somepath/devilutionx/saves/devilutionx --ini-path /somepath/devilution/ini/devx-1.0.3.ini
/somepath/devilutionx-1.0.3 -n -x --save-dir /somepath/devilutionx/saves/mule1 --ini-path /somepath/devilution/ini/devx-1.0.3-mules.ini

/somepath/devilutionx-1.1.0 -n  --save-dir /somepath/devilutionx/saves/devilutionx --ini-path /somepath/devilution/ini/devx-1.1.0.ini
/somepath/devilutionx-1.1.0 -n -x --save-dir /somepath/devilutionx/saves/mule1 --ini-path /somepath/devilution/ini/devx-1.1.0-mules.ini

**Comment by qndel** [comment]
`printInConsole("    %-20s %-30s\n", "--config-dir", "Specify the location of diablo.ini");`

**Reply by Trihedraf** [reply]
That will only work in 1.2.0 and later. Because it wasn’t added until after 1.1.0

**Comment by AJenbo** [comment, marked as answer]
In 1.2.0 all the ini settings are renamed so in one way they could coexist in the same file while having different settings. We haven't had much interest in the 1.0.x so maintenance of it is of low priority. I did backport 4/9 of the 1.2.0 bug fixes. I can try and see if it would be simple to also backport the `--config-dir` feature and some more of the bug fixes, but no promises :)

**Reply by mgpat-gm** [reply]
If the latest release has the --config-dir, that satisfies my query/feature request, as I can leave the 1.0.3 config files where they are and just direct the latest release to a new config location.

**Reply by AJenbo** [reply]
That's also valid, and yes it does have --config-dir :)

**Reply by qqkookie** [reply]
IMHO, it would be nice to have direct `diablo.ini` config file path, instead of separate config-dir folder There are not many config ini files to warrant separate config directory, just one, `diablo.ini` file. So, if user specifies `--config  my-config.ini`, then it will search `my-config.ini` in save directory (default save directory or specified by `--save-dir`) or diablo.ini file.

And as related topic, I think that default save directory (`%APPDATA%\Roaming\diasurgical\devilution`) is not so obvious default save/config directory. It is too cryptic. I prefer `SAVE` folder in the game directory as default save directory choice as many plain user expects. It will also help to avoid problem of user with international character in user name (Read-Only Directory Error #1862).

**Reply by StephenCWills** [reply]
Writing to the `AppData` folder is standard practice for applications based on security practices implemented by Microsoft starting with Windows Vista. Attempting to write to the game folder causes a ton of issues because the default security settings disallow writes to anything under `Program Files` without explicitly elevated privileges. Furthermore, encouraging users to bypass the default security settings or run the game using elevated permissions is a somewhat common and very bad practice so I appreciate the developers on this project for not doing that.

In other words, you can blame Microsoft for the cryptic savegame location. I think it would be a terrible idea to change it. However, the `--config` option does sound like a pretty good idea to me.

**Reply by AJenbo** [reply]
The zerotier configuration files also go in the config folder, other similar settings files may come later. if you do not provide --config-dir it will use the same path as --save-dir.

Some users install the game in there home folder, and so the internattional symbols issue would still apply in these cases. Besides it's not a good solution for what is really a bug (that has been fixed). What if a Spanish speaker installs it to c:/Díablo.

**Reply by qqkookie** [reply]
> Writing to the `AppData` folder is standard practice for applications based on security practices implemented by Microsoft starting with Windows Vista. Attempting to write to the game folder causes a ton of issues because the default security settings disallow writes to anything under `Program Files` without explicitly elevated privileges. Furthermore, encouraging users to bypass the default security settings or run the game using elevated permissions is a somewhat common and very bad practice so I appreciate the developers on this project for not doing that.

The MS security recommendation applies only to program that is installed by admin right.  This program is not  **installed** by admin with setup program in public `C:\Program Files`, which need admin right to write. Most people (non-admin) will just simply unzip distribution zip file in their personal folder and start play.  So such security hazard does not exist. Many non-installation game (unzip and play, so-called portable distribution) generally use the unzipped game directory as save directory without any security hazard to keep all data to local in game directory.

So my suggestion is like this:  If game is installed by admin in directory which is not writable by non-privileged user like `C:\Program Files`, then use `%APPDATA%\devilution` folder as save/config directory. Else, if game is just unzipped and play, then use unzipped game directory\SAVE as default save/config folder. This will not make any security weakness.

> The zerotier configuration files also go in the config folder, other similar settings files may come later. if you do not provide --config-dir it will use the same path as --save-dir.

Multiple config files for a game? It would confuse player. How about add new `[Section]` in `diablo.ini`?  And version 1.2 does not honour  the `--save-dir` when searching for config file `diablo.ini`. To change both, both  --save-dir ./SAVE' and  --config-dir ./SAVE'  are needed, which seems awkward.  To keep it simple, please keep all config infos in single config file as separate ini section and no separate config directory. 

> Some users install the game in there home folder, and so the internattional symbols issue would still apply in these cases. Besides it's not a good solution for what is really a bug (that has been fixed). What if a Spanish speaker installs it to c:/Díablo.

Opening save/config file with **relative path** to current directory where game is executed and process is running (not absolute full path) will solve the problem. Then,` c:/Díablo` or  international user name will not cause problem.

**Reply by StephenCWills** [reply]
> So my suggestion is like this: If game is installed by admin in directory which is not writable by non-privileged user like `C:\Program Files`, then use `%APPDATA%\devilution` folder as save/config directory. Else, if game is just unzipped and play, then use unzipped game directory\SAVE as default save/config folder. This will not make any security weakness.

You should have a look at this PR.
https://github.com/diasurgical/devilutionX/pull/1736

**Reply by qqkookie** [reply]
I will explain why save file in `C:\Program Files\DeviolutionX\SAVE` is bad practice and MS is against such. In such case, game should have write permission on the SAVE folder, where admin can only write, to save the game. So game must be run with admin privilege elevation, which is dangerous security hazard. So MS prohibits such practice.
  
And if the game is installed in shared `C:\Program Files\DevilutionX` folder and played by multiple user, then `C:\Program Files\DevilutionX\SAVE` is obviously bad choice, which result in conflicting character save file (single_0.sv) among players and can not support multi-user environment. So it is needed to use  `%APPDATA%\devilution` to support multi-user environment. That is why MS recommends  use of `%APPDATA%` folder.

But if the game is simply unzipped in personal folder like  `C:\Users\<Username>\DevilutionX` by non-admin user  and executed without admin privilege elevation, game can write save in  `C:\Users\<Username>\DevilutionX\SAVE`   safely. So it poses no security hazard regarding admin permission/right or multi-user conflict. The `deviltionx.exe` does not require admin privilege elevation and the SAVE folder is not shared among multiple users, so it is safe to write and save in personal SAVE`  directory.  It will not cause save file conflict among multiple users. 

This is why I suggested `%APPDATA%\devilution`  for shared multi-user  environment, and `$game-directory\SAVE`  for unzipped into  personal folder case.

**Reply by AJenbo** [reply]
All of this may or may not be true depending on user action and may change in the future (we may add an installer to the Windows releases). Detecting things like this in a cross platform way is non-trivial, so we rely on SDL to sort this out for the most part. 

> Multiple config files for a game? It would confuse player. How about add new [Section] in diablo.ini?

I don't think you realize what ZeroTier is or what it's configuration data looks like.

**Comment by julealgon** [comment]
@mgpat-gm 
> I mainly play versin 1.0.3.

Why though?

**Reply by mgpat-gm** [reply]
I prefer playing the game close to the original without the "Bugfixes for original Diablo bugs" that effect gameplay. Some are minor or not an issue, but others are fairly significant.

That said, one day I hope to be able to have item highlight and a stash :)

**Reply by AJenbo** [reply]
You will have to port it your self, or some one else will have to, it's not going to land in 1.0.x :/

**Reply by mgpat-gm** [reply]
I know :frowning_face: I'm fairly content just being able to play on linux. I'll see if I can figure out git a little better, merge what I can and cross my fingers. :smile:

**Comment by qqkookie** [comment]
In other thread #1804, I suggested using `Diablo.ini` entry for rarely used command line options.
So I suggest to reduce number of user facing command line options.
Most debug options are moved to `[Debug]` section  of `Diablo.ini`. 
```
--h, --help, --version: preserved 
-w  windowed mode
-x  no intro video
--diablo force diablo 
--config <file.ini>
--user-dir [<dir>]  ; replace --save-dir + --config-dir

Debug options
-m <m>   ; preserved
```
If `--user-dir` is used without any parameter, it will default to  ${PreferenceDir}(Or %APPDATA%)/devilution.
if `--user-dir` is used with directory, then  use dir as both `--saved-dir` and `--config-dir Diablo.ini`
If both `--user-dir <dir>` and `--config <file.ini>`  are used, then look `<dir>/<file.ini>`.

These rarely used command line options below may be moved to `Diablo.ini` `[section]` entries.
```
--data-dir   --> [Global] Mpq Dir 
--lang-dir   --> [Global] Language
--ttf-dir, --ttf-name --> [Global] Font
--spawn      --> [Global] Force Spawn
--vanilla    --> [Global] OrignalSave
--verbose    --> [Global] Log
-f           --> [Grahics] Show FPS
--nestart    --> [Hellfire] Alt NestArt
--config-dir --> [Network] ZeroTierConfig
--diablo     --> [Global] Force Diablo, but override with --diablo option 
-w           --> [Graphics] FullScreen, but override with -w option
-x           --> [Global] Intro, but override with -x option 

[Debug] section
Cheat Mode=0      ; -w
God Mode=0        ; -$
Debug Tool=0      ; -^
View Visibility=0 ; -v
Net Timeout=0     ; -i
Trigger Level=0   ; -j <l>
Dungeon=0 0       ; -l <l> <t>
Force Quest=0     ; -q <q>
Debug_S           ; -s
Map Seed=0        ; -r <n>
Quest Level=0     ; -t <q>
```
So the suggested `Diablo.ini` improvements in #1804 are implemented, these rarely used command line options may be removed and replaced with `Diablo.ini` entry.

**Reply by AJenbo** [reply]
The arguments are not intended to be configurations

**Reply by qqkookie** [reply]
Indeed, they are not intended to be configured in in-game menu. So in-game menu can simply ignore them. 

But command console and command line is inherently not user friendly for Windows and other modern platform user. On most platforms and most non-Linux or non-shell user, symptom of  too many command line options (of most Linux centric command) is too painful to use such command line options. Even typing or opening console is challenging in some platform without physical keyboard (e.g. smartphone) . Most ordinary player abhors even opening the command console. It should be reserved as last resort measure, not primary way to use such options. 

I think that `Diablo.ini` is good place to provide alternate way to save such long and tedious typing of these multitude of options in command line through keyboard.
And entries in `Diablo.ini`  is self-explanatory and easier to understand . The "Show FPS=1" line in `Diablo.ini` is easier than figuring out `-f` for FPS.

**Reply by AJenbo** [reply]
Even Steam as an option for entering command arguments, i think it can be considered pretty common place.

**Reply by qqkookie** [reply]
Having such command line option is not bad thing. But most windows game does not use or expect for player to use it as primary way to setting player game options. The very existence of such "launcher/helper" program shows how command line option is painful for ordinary player to use. Many modern platform (e.g. smartphone) does not have such game launcher like steam client that user can set command line options for individual games. 

Primary way is mostly in-game, on-screen option setting page and such options are saved in some configuration file. For some detailed or rarely used options that are not handled/exposed by such option page can be modified by editing such configuration file. It is also common practice. I think that having such alternative way is not bad thing, too.

**Reply by AJenbo** [reply]
They are not intended for players to use.

**Reply by julealgon** [reply]
> But command console and command line is inherently not user friendly for Windows

The commands that are toggleable via cmdline are not intended to be user friendly in the first place: they are advanced commands.
